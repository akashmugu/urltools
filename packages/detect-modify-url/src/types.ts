import { DetectModifyUrlConfig, SiteConfig } from '@urltools/utils/lib/types/detectModifyUrl'

export type ModifiedUrl = {
  url: string
  match?: SiteConfig
}

export type DetectModifyUrl = (config: DetectModifyUrlConfig) => (url: string) => ModifiedUrl
