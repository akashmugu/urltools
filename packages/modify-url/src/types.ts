import { ModifyUrlConfig } from '@urltools/utils/lib/types/modifyUrl'

export type ModifyUrl = (config: ModifyUrlConfig) => (url: string) => string

export type ModifyUrlSubConfig = Exclude<ModifyUrlConfig[keyof ModifyUrlConfig], undefined>

export type Module = (url: URL, config: ModifyUrlSubConfig) => URL
