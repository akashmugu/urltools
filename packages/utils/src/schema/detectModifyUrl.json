{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "detectModifyUrlConfig",
  "type": "array",
  "items": {
    "title": "siteConfig",
    "type": "object",
    "properties": {
      "name": {
        "type": "string",
        "examples": ["Google Search"]
      },
      "test": {
        "type": "string",
        "examples": ["^(?:https|http)://(?:www.)?google.com/search?"]
      },
      "filters": {
        "type": "array",
        "items": {
          "title": "modifyUrlConfig",
          "type": "object",
          "examples": [{ "searchParams": { "type": "pick", "value": ["q"] } }]
        }
      }
    },
    "required": ["name", "test", "filters"],
    "additionalProperties": false
  }
}
