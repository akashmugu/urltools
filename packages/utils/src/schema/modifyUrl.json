{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "modifyUrl.json",
  "title": "modifyUrlConfig",
  "type": "object",
  "properties": {
    "protocol": {
      "title": "protocol",
      "type": "string"
    },
    "username": {
      "title": "username",
      "type": "string"
    },
    "password": {
      "title": "password",
      "type": "string"
    },
    "host": {
      "title": "host",
      "type": "string"
    },
    "pathname": {
      "title": "pathname",
      "type": "object",
      "properties": {
        "type": {
          "type": "string",
          "enum": ["replace", "append"]
        },
        "value": {
          "type": "string"
        }
      },
      "required": ["type", "value"],
      "additionalProperties": false
    },
    "searchParams": {
      "oneOf": [
        {
          "title": "searchParamsPickOmit",
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "enum": ["pick", "omit"]
            },
            "value": {
              "type": "array",
              "items": { "type": "string" }
            }
          },
          "required": ["type", "value"],
          "additionalProperties": false
        },
        {
          "title": "searchParamsReplaceAppend",
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "enum": ["replace", "append"]
            },
            "value": {
              "type": "object"
            }
          },
          "required": ["type", "value"],
          "additionalProperties": false
        }
      ]
    },
    "hash": {
      "title": "hash",
      "type": "string"
    }
  },
  "additionalProperties": false
}
