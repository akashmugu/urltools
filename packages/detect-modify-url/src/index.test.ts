import { DetectModifyUrlConfig } from '@urltools/utils/lib/types/detectModifyUrl'
import { ModifyUrlConfig } from '@urltools/utils/lib/types/modifyUrl'
import detectModifyUrl from './index'

describe('detectModifyUrl', () => {
  it('throws invalid config', () => {
    expect(() => {
      // @ts-ignore
      detectModifyUrl({ foo: 'bar' })
    }).toThrowErrorMatchingSnapshot()
  })

  it('returns fn for valid config', () => {
    expect(detectModifyUrl([]) instanceof Function).toBe(true)
  })

  describe('no match', () => {
    test('no sites', () => {
      const res = detectModifyUrl([])('http://www.google.com/search?foo=bar')
      expect(res.url).toBe('http://www.google.com/search?foo=bar')
      expect(res.match).toBe(undefined)
    })

    test('1 site', () => {
      const config: DetectModifyUrlConfig = [
        {
          name: 'site 1',
          test: '^(?:https|http)://(?:www.)?site1.com',
          filters: []
        }
      ]

      const res = detectModifyUrl(config)('http://www.site2.com')
      expect(res.url).toBe('http://www.site2.com')
      expect(res.match).toBe(undefined)
    })

    test('multiple site', () => {
      const config: DetectModifyUrlConfig = [
        {
          name: 'site 1',
          test: '^(?:https|http)://(?:www.)?site1.com',
          filters: []
        },
        {
          name: 'site 2',
          test: '^(?:https|http)://(?:www.)?site2.com',
          filters: []
        }
      ]

      const res = detectModifyUrl(config)('http://www.site3.com')
      expect(res.url).toBe('http://www.site3.com')
      expect(res.match).toBe(undefined)
    })
  })

  describe('match', () => {
    test('1 site - no filters', () => {
      const config: DetectModifyUrlConfig = [
        {
          name: 'site 1',
          test: '^(?:https|http)://(?:www.)?site1.com',
          filters: []
        }
      ]

      const res = detectModifyUrl(config)('http://www.site1.com/?q=galaxy&foo=bar')
      expect(res.url).toBe('http://www.site1.com/?q=galaxy&foo=bar')
      expect(res.match).toBe(config[0])
    })

    test('1 site - 1 filter', () => {
      const subconfig: ModifyUrlConfig = {
        searchParams: { type: 'pick', value: ['q'] }
      }
      const config: DetectModifyUrlConfig = [
        {
          name: 'site 1',
          test: '^(?:https|http)://(?:www.)?site1.com',
          filters: [subconfig]
        }
      ]

      const res = detectModifyUrl(config)('http://www.site1.com/?q=galaxy&foo=bar')
      expect(res.url).toBe('http://www.site1.com/?q=galaxy')
      expect(res.match).toBe(config[0])
    })

    test('1 site - multiple filters', () => {
      const subconfig1: ModifyUrlConfig = {
        searchParams: { type: 'pick', value: ['q'] }
      }
      const subconfig2: ModifyUrlConfig = {
        searchParams: { type: 'append', value: { name: 'akash' } }
      }
      const config: DetectModifyUrlConfig = [
        {
          name: 'google search',
          test: '^(?:https|http)://(?:www.)?google.com/search?',
          filters: [subconfig1, subconfig2]
        }
      ]

      const res = detectModifyUrl(config)('http://google.com/search?q=galaxy&undesirablefoo=bar&pryingtrackid=baz')
      expect(res.url).toBe('http://google.com/search?q=galaxy&name=akash')
      expect(res.match).toBe(config[0])
    })

    test('multiple sites', () => {
      const subconfig1: ModifyUrlConfig = {
        searchParams: { type: 'pick', value: ['q'] }
      }
      const subconfig2: ModifyUrlConfig = {
        searchParams: { type: 'pick', value: ['foo'] }
      }
      const config: DetectModifyUrlConfig = [
        {
          name: 'google search',
          test: '^(?:https|http)://(?:www.)?google.com/search?',
          filters: [subconfig1]
        },
        {
          name: 'yahoo search',
          test: '^(?:https|http)://(?:www.)?yahoo.com/search?',
          filters: [subconfig2]
        }
      ]

      const processUrl = detectModifyUrl(config)

      const res1 = processUrl('http://google.com/search?q=galaxy&foo=bar&pryingtrackid=baz')
      expect(res1.url).toBe('http://google.com/search?q=galaxy')
      expect(res1.match).toBe(config[0])

      const res2 = processUrl('http://yahoo.com/search?q=galaxy&foo=bar&pryingtrackid=baz')
      expect(res2.url).toBe('http://yahoo.com/search?foo=bar')
      expect(res2.match).toBe(config[1])
    })
  })
})
