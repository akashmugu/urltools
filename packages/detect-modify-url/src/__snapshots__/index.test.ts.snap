// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detectModifyUrl throws invalid config 1`] = `"[{\\"keyword\\":\\"type\\",\\"dataPath\\":\\"\\",\\"schemaPath\\":\\"#/type\\",\\"params\\":{\\"type\\":\\"array\\"},\\"message\\":\\"should be array\\"}]"`;
